Hardware id and release counters

shikhar394 opened the pull request:
Greetings, 
I have implemented release counters and hardwareIDs in all the relevant Uptane modules, including errors and exceptions linked to it. 
I have also demonstrated 3 attacks:

- Image_rollback_attack : A compromised director attempting to install an image with an outdated release counter.
- Confused_bundle_attack : A compromised director attempting to install an image with mismatching release counters.
- Sneaky_director_attack : A compromised director attempting to install an image with mismatching hardware IDs. Basically, intending to install an image on an ECU which it was not meant for through leveraging the ECU Serial. 

All these attacks can be called in the script through the following, please see the function definitions for more advanced function calls: 
The names correspond with the usual demos.
**image_rollback_attack**
di.image_rollback_attack("Attack1.img")
dd.image_rollback_attack("Attack1.img")
dp.update_cycle()
ds.update_cycle()

**confused_bundle_attack**
di.confused_bundle_attack("Attack2.img")
dd.confused_bundle_attack("Attack2.img")
dp.update_cycle()
ds.update_cycle()

**sneaky_director_attack**
di.sneaky_director_attack("Attack3.img")
dd.sneaky_director_attack("Attack3.img")
dp.update_cycle()

I have also added a couple of launch_* files that are just python scripts to launch the demo
use:
python -i launch_[name].py

shikhar394 committed:
Added test for test_primary in __init__ for invalid director name.

shikhar394 committed:
Tested 1) Invalid names for director_repo_name in __init__() 2) Re-registering a secondary with a primary in register_new_secondary() 3) update_exists_for_ecu()

shikhar394 committed:
Removed primary update cycle

shikhar394 committed:
Code not working: bad signature

shikhar394 committed:
File Error, tracing to fix it

shikhar394 committed:
Deep test version for update_exists_for_ecu()

shikhar394 committed:
Working Deep test version for update_exists_for_ecu()

shikhar394 committed:
Not removing pinned.json so other pieces of code that use it can still work

shikhar394 committed:
Added support for Python2 to specify encoding

shikhar394 committed:
Working commit

shikhar394 committed:
added docstring for hardwareID and release counter

shikhar394 committed:
Included check for hardwareID and release counter in primary [still need to see how primaries install updates]

shikhar394 committed:
Made release_counter a dictionary and changed primary, secondary and formats to reflect changes

shikhar394 committed:
pushing changes

shikhar394 committed:
Waiting on making asn1 work

awwad committed:
Support Release Counter and Hardware ID in services

shikhar394 committed:
Primary's clean_slate() works

shikhar394 committed:
Working Primary and demo_primary with manifests being submitted

shikhar394 committed:
Working secondary files

shikhar394 committed:
Working_releaseCounters_and_hardware_ID

shikhar394 committed:
Fixed edge case in the imagerepos and director repos comparison

awwad commented on pinned.json:
```diff
@@ -0,0 +1 @@
+{"delegations":[{"paths":["*"],"repositories":["imagerepo","director"]}],"repositories":{"director":{"mirrors":["file:////home/shikhar/Desktop/uptane/samples/metadata_samples_long_expiry/update_to_one_ecu/full_metadata_archive/director/"]},"imagerepo":{"mirrors":["file:////home/shikhar/Desktop/uptane/samples/metadata_samples_long_expiry/update_to_one_ecu/full_metadata_archive/director/"]}}}
```
Is this particular pinned.json file necessary? This shouldn't be in the root directory if so, and it shouldn't reference `file:///home/shikhar/Desktop/etc`. Was it added to git by mistake?

awwad commented on tests/test_data/pinned.json:
```diff
@@ -1,16 +1 @@
-  ]
-}
\ No newline at end of file
+{"delegations":[{"paths":["*"],"repositories":["imagerepo","director"]}],"repositories":{"director":{"mirrors":["file:////home/shikhar/Desktop/uptane/tests/test_data/temp_test_primary/director"]},"imagerepo":{"mirrors":["file:////home/shikhar/Desktop/uptane/tests/test_data/temp_test_primary/imagerepo"]}}}
```
This was generic and readable content used in tests. It is now specific to a particular machine, referencing `file:////home/shikhar/Desktop/uptane`. If this is now a dynamically generated file that is always produced before it is needed, then it should be removed from git entirely. Is that the case? Or does it need to be set manually in order for tests to work? (That would not do.) Or has it been replaced by tests/test_data/pinned_template.json? (In that case as well, it should be removed from git.)

shikhar394 replied:
Yep, totally agree. I just remembered it was the test_secondary that needed a pinned.json file to function (pointed by travis CI). Working on making test_secondary dynamic right now.
Thanks!

awwad commented on tests/test_primary.py:
```diff
@@ -13,7 +13,7 @@
 
 import uptane # Import before TUF modules; may change tuf.conf values.
-
+from io import open
```
Opinions on this may differ, but I think you should add a brief comment explaining why this is necessary. (Python2 compatibility providing encoding in an open() call? Or what?)

awwad commented on tests/test_primary.py:
```diff
@@ -35,6 +35,9 @@
 import demo # for generate_key, import_public_key, import_private_key
+import json
+import canonicaljson
+import codecs
```
codecs is not used and so this import should be removed.

awwad commented on uptane/__init__.py:
```diff
@@ -12,7 +12,7 @@
  # Configure TUF to use DER format instead of Python dictionaries / JSON.
 import tuf.conf
-tuf.conf.METADATA_FORMAT = 'der'
+tuf.conf.METADATA_FORMAT = 'json'
```
The default encoding for Uptane should not be changed. Not all modifications made locally should be committed to git. (I know that hardware ID and release counter probably don't work in ASN.1/DER yet, but this value still shouldn't change.)

awwad commented on uptane/__init__.py:
```diff
@@ -44,6 +44,25 @@ class Spoofing(Error):
+  """
+  pass
+
+class BadReleaseCounterValue(Error):
```
I would call this ImageRollback or something along those lines.

awwad commented on uptane/__init__.py:
```diff
@@ -44,6 +44,25 @@ class Spoofing(Error):
+  pass
+
+
+class BadReleaseCounterValue(Error):
```
This error class is duplicated. (Accident?)

awwad commented on uptane/clients/primary.py:
```diff
@@ -38,7 +38,7 @@
 import uptane.services.timeserver as timeserver
 import uptane.encoding.asn1_codec as asn1_codec
-
+from pprint import pprint
```
Debug statements shouldn't be in PRs. (I advise leaving them out of git entirely.)

awwad commented on uptane/clients/primary.py:
```diff
@@ -404,9 +417,14 @@ def get_validated_target_info(self, target_filepath):
         target_filepath, multi_custom=True)
+    #print("TARGET", target_filepath)
+    #print("TARGET_INFO")
+    #pprint(validated_target_info)
```
As noted above, debug statements shouldn't be in PRs. (I advise leaving them out of git entirely.)

awwad commented on uptane/clients/primary.py:
```diff
@@ -221,6 +226,8 @@ def __init__(
     primary_key,
     time,
     timeserver_public_key,
+    hardware_id = "Infotainment101",   
```
You shouldn't have a default hardware ID in the reference implementation. You can have one in the demo, but the reference implementation should be clean and simply require hardware_id as an input argument to the `__init__` call.

That aside, we're not going to be using the Primary's hardware_id or release_counter yet; that will come into play when we add Secondary functionality to the Primary later, as discussed above and below. It's okay to leave these two arguments in the constructor, since those values will eventually come in this way.

awwad commented on uptane/clients/primary.py:
```diff
@@ -221,6 +226,8 @@ def __init__(
     time,
     timeserver_public_key,
+    hardware_id = "Infotainment101",   
+    release_counter = 1,  # Sample image has version 0
```
For release_counter, a default is probably OK, but the default should be 0. (The Implementation Specification describes that.)

awwad commented on uptane/clients/primary.py:
```diff
@@ -315,7 +326,9 @@ def __init__(
-
+  def __str__(self):
+    a = "VIN: {}, ECU_SERIAL: {}, HARDWARE_ID: {}, RELEASE_COUNTER: {}".format(self.vin, self.ecu_serial, self.hardware_id, self.release_counter)
+    return a
```
`__str__` is probably handy. In any case, though:
- Don't use the temp variable; just return these values.
- Don't exceed 80 columns. See [this lab guideline](https://github.com/secure-systems-lab/code-style-guidelines#maximum-line-length).

Try for example:
```Python
  return "VIN: ........ ".format(...,
      ...)
```

awwad commented on uptane/clients/primary.py:
```diff
@@ -404,9 +417,14 @@ def get_validated_target_info(self, target_filepath):
     tuf.formats.RELPATH_SCHEMA.check_match(target_filepath)
-
+    Hardware_ID_Release_Counter = set() # To check if the hardwareIDs and release counters are matching in all the repos other than director
+    Director_HardwareID_Release_counter = set() # To check with that of the other repos
```
You don't need these variables here, nor would you need them to be sets, and you should observe the same capitalization customs for variables used elsewhere in the code (all lowercase; [lab guideline here](https://github.com/secure-systems-lab/code-style-guidelines#general-variable-names))

You'll need temp variables down around right-462, but I think you'll just need two flat values: one release counter integer and one hardware id string. See comments below.

awwad commented on uptane/clients/primary.py:
```diff
@@ -449,9 +512,6 @@ def get_validated_target_info(self, target_filepath):
 
-
-
-
```
The Uptane reference implementation uses 5 blank lines between methods. NBD, but no need to delete these lines.

shikhar394 replied:
This is strange that this issue hasn't collapsed yet! I fixed it long back with the rest of the fixes and my code log shows that my updated code (local) is up-to-date with the one online. I'll push it back when we resolve the merge conflicts.

awwad commented on uptane/clients/primary.py:
```diff
@@ -564,22 +625,47 @@ def primary_update_cycle(self):
             'fileinfo!' + ENDCOLORS)
         continue
 
+      # Checking the target info for the primary itself. Disregarding if the hardware ID or release counters do not satisfy the requirements
```
This Primary class is not set up to process updates for itself, so this code doesn't make sense right now. (Please exclude this section.) In order to update a Primary ECU itself, rather than duplicating code we'll either include an instance of a Secondary object in the Primary class, or have class Primary inherit from class Secondary or something along those lines, depending on which is least confusing to a reader. After that, we'll tweak the Director/InventoryDB slightly to accommodate the change.

Also, lines should not exceed 80 characters.

awwad commented on uptane/clients/primary.py:
```diff
@@ -855,12 +941,15 @@ def generate_signed_vehicle_manifest(self):
     #     'signed': vehicle_manifest,
     #     'signatures': []
     # }
+    #print("vehicle manifest", vehicle_manifest)
```
Please remove all debug statements.

awwad commented on uptane/clients/primary.py:
```diff
@@ -932,7 +1019,7 @@ def _check_ecu_serial(self, ecu_serial):
 
   def register_ecu_manifest(
-      self, vin, ecu_serial, nonce, signed_ecu_manifest, force_pydict=False):
+      self, vin, ecu_serial, hardware_id, release_counter, nonce, signed_ecu_manifest, force_pydict=False):
```
Do these need to be here for some reason? I don't think the Primary needs to know a Secondary's hardware_id or release_counter - except in passing on the ecu manifest itself. The hardware_id and release_counter are in the signed ECU manifest itself (which looks to be correct later on).

awwad commented on uptane/clients/secondary.py:
```diff
@@ -41,7 +42,7 @@
 
 from uptane import GREEN, RED, YELLOW, ENDCOLORS
-
+from demo.uptane_banners import *
```
Why are we adding more demo code into the reference implementation? (I can think of one reason this may be reasonable to do in the short-term.) What are you using?

awwad commented on uptane/clients/secondary.py:
```diff
@@ -73,6 +74,13 @@ class Secondary(object):
       unique.) The Director should be aware of this identifier.
 
+    self.hardware_id
+      A unique identifier for an ECU through it's hardware ID. Conforms to uptane.formats.HARDWARE_ID_SCHEMA. This is used to prevent a compromised director from causing an ECU to download an image not intended for it. 
```
`hardware_id` does not uniquely identify an ECU; `ecu_serial` should. `hardware_id` may uniquely identify a group of ECUs. We ensure that firmware installed on any of those ECUs may only be firmware that was indicated to be OK to use on that group of ECUs by the Image Repository and Director. We should be very clear about this here.

Also, line length.

awwad commented on uptane/clients/secondary.py:
```diff
@@ -73,6 +74,13 @@ class Secondary(object):
+
+
+    self.release_counter
+      A dictionary wih counters to track the version number of the images installed. Conforms to uptane.formats.RELEASE_COUNTER_SCHEMA. This is used to prevent a compromised director from causing an ECU to download an outdated image or an older one with known vulnerabilities.  
```
Is release_counter still a dictionary? We should update this comment otherwise.

"outdated" may be misleading; should consider something like "an older image than the one installed", which may also be slightly misleading, but less so.... (Technically, use of release_counter simply avoids an image being installed if the new image has a lower release counter than that of the image currently marked as installed on this ECU. There are edge cases where new images might have lower release counters, if the firmware producers intend them as new intermediate upgrades, or any number of other odd cases... but "an older image than the one installed" is probably fine.)

Also, line length.

awwad commented on uptane/clients/secondary.py:
```diff
@@ -173,6 +181,8 @@ def __init__(
     ecu_key,
     time,
     timeserver_public_key,
+    release_counter = 1,
```
Default 0.

awwad commented on uptane/clients/secondary.py:
```diff
@@ -173,6 +181,8 @@ def __init__(
     time,
     timeserver_public_key,
+    release_counter = 1,
+    hardware_id = "SecondaryPotato101",
```
Require a value for `hardware_id` rather than allowing a default and using a nonsense default. (That will also require you to swap the order of these arguments, since you can't have a required arg after an optional arg.)

Incidentally, even in demo code (which this isn't anyway), we shouldn't use silly names. (:

awwad commented on uptane/clients/secondary.py:
```diff
@@ -274,7 +288,10 @@ def __init__(
-
+  def __str__(self):
+    a = "VIN: {}, ECU_SERIAL: {}, HARDWARE_ID: {}, RELEASE_COUNTER: {}".format(self.vin, self.ecu_serial, self.hardware_id, self.release_counter)
+    #print("MYSELF")
```
As before:
- line length
- debug statement
- unnecessary temp variable

awwad commented on uptane/clients/secondary.py:
```diff
@@ -483,20 +502,34 @@ def fully_validate_metadata(self):
     for target in self.updater.targets_of_role(
         rolename='targets', repo_name=self.director_repo_name):
 
-      # Ignore target info not marked as being for this ECU.
```
Why remove this explanatory comment? (Is it misleading or something?)

shikhar394 replied:
My bad, I must be changing some part of the code and might have deleted the comment by mistake.

awwad commented on uptane/clients/secondary.py:
```diff
@@ -483,20 +502,34 @@ def fully_validate_metadata(self):
         continue
 
+      else:
+        if self.hardware_id != target['fileinfo']['custom']['hardware_id']:
```
This can be an elif instead of a nested else if. Also, line length in the line below.

awwad commented on uptane/clients/secondary.py:
```diff
@@ -483,20 +502,34 @@ def fully_validate_metadata(self):
+          log.warning(RED + 'Received a target from the Director with instructions to install an Image on self with ECU_Serial {} with mismatching hardwareID. Diregarding/not downloading target for saving. The target is {}'.format(self.ecu_serial, repr(target))+ ENDCOLORS)
+          continue
+
+        if self.release_counter > target['fileinfo']['custom']['release_counter']:
```
This can be another elif at the previous level rather than a nested if.

awwad commented on uptane/clients/secondary.py:
```diff
@@ -483,20 +502,34 @@ def fully_validate_metadata(self):
+
+        if self.release_counter > target['fileinfo']['custom']['release_counter']:
+          log.warning(RED + 'Received a target from the Director with instructions to install an Image {} on self with ECU_Serial {} with lower value of release counter than current. Diregarding/not downloading target for saving. The target is {}'.format(target['filepath'], self.ecu_serial, repr(target))+ ENDCOLORS)
+          raise uptane.BadReleaseCounterValue("The director has instructed the ECU to download an image that has a lower relase counter than the current. Image rejected")
```
It's probably worth noting the two release_counter values.

awwad commented on uptane/services/director.py:
```diff
@@ -173,7 +173,20 @@ def validate_ecu_manifest(self, ecu_serial, signed_ecu_manifest):
           'origin ECU (' + repr(ecu_serial) + ') is not the same as what is '
           'signed in the manifest itself (' +
           repr(signed_ecu_manifest['signed']['ecu_serial']) + ').')
+    """
```
Since the Primary doesn't need to know about these two values for its Secondaries, these lines can be excluded.

awwad commented on uptane/services/director.py:
```diff
@@ -265,7 +278,10 @@ def register_vehicle_manifest(
+      #print("Signed vehicle manifest", signed_vehicle_manifest)
 
+    #print(signed_vehicle_manifest)
+    
```
debug lines

awwad commented on uptane/services/director.py:
```diff
@@ -293,7 +309,7 @@ def register_vehicle_manifest(
     all_ecu_manifests = \
         signed_vehicle_manifest['signed']['ecu_version_manifests']
-
+   
```
trailing whitespace

awwad commented on uptane/clients/primary.py:
```diff
@@ -690,12 +776,12 @@ def primary_update_cycle(self):
     # for partial-verifying Secondaries. In both cases, the files are swapped
     # into place atomically after being constructed or copied. Secondaries
     # may be requesting these files live.
+    if new_self_release_counter != -1:
```
These lines are unnecessary, I think. See above.

awwad commented on uptane/clients/primary.py:
```diff
@@ -479,6 +539,7 @@ def primary_update_cycle(self):
     """
     log.debug('Refreshing top level metadata from all repositories.')
     self.refresh_toplevel_metadata_from_repositories()
+    new_self_release_counter = -1
```
Can be removed. See comments below.

awwad commented on uptane/clients/primary.py:
```diff
@@ -564,22 +625,47 @@ def primary_update_cycle(self):
+        for custom_field in ['ecu_serial', 'hardware_id', 'release_counter']:
+          if custom_field not in targetinfo['fileinfo']['custom']:
+            raise uptane.Error(
+              "Director repo failed to include an {} for a target. Target metadata was: {}".format(custom_field, repr(target)))
```
I don't think these changes are useful. The Primary needs to know the ECU Serial in order to distribute the image to the Secondaries, which is why that is checked. The Primary doesn't need to know the hardware_id or release_counter currently installed for each Secondary or of the firmware to be distributed to each secondary, and is not a good authority for that information.

I refer here to changes appearing from left-567 to left-693 (right-627 to right-782).

shikhar394 replied:
I completely agree. I did this to make checks in the update cycle for the primary (I thought the primary updates worked). 

Is it fine if I keep this till we decide on how to implement the primary's updates? Or would you prefer if I removed these anyway?

shikhar394 replied:
Just read the following comments, I'll make the changes. Thanks!

awwad commented on uptane/clients/primary.py:
```diff
@@ -77,6 +77,12 @@ class Primary(object): # Consider inheriting from Secondary and refactoring.
       (In other implementations, the important point is that this should be
       unique.) The Director should be aware of this identifier.
 
+    self.hardware_id
```
Class Primary doesn't track updates for itself. The way it will do this will be either to inherit from Secondary or to include a Secondary object in each instance for itself, rather than to duplicate code, so all code added to this module specifically for the purpose of testing the Primary's own updates should be removed (as it can't be functional yet in any case). So these fields are not necessary. See further comments below for more detail.

awwad commented on uptane/clients/primary.py:
```diff
@@ -266,13 +273,17 @@ def __init__(
     self.vin = vin
     self.ecu_serial = ecu_serial
+    self.hardware_id = hardware_id
+    self.release_counter = release_counter
```
These lines are fine, but be aware that real use of the release_counter and hardware_id for the Primary itself will follow later.

awwad commented on uptane/clients/primary.py:
```diff
@@ -441,6 +459,51 @@ def get_validated_target_info(self, target_filepath):
+
+
+
+
```
Please kill the trailing whitespace in the lines added above. You use Sublime, so I recommend the Sublime package TrailingSpaces, which will pester you when you leave trailing whitespace.

awwad commented on uptane/clients/primary.py:
```diff
@@ -441,6 +459,51 @@ def get_validated_target_info(self, target_filepath):
+
+
+
+
```
I may have missed something, but I don't think this code needs to be as complicated as it is. Your goal is to make sure that all repositories provide a hardware_id and release_counter and that they all match. You shouldn't need to distinguish between Director and Image Repository/ies here. Please simplify.

shikhar394 replied:
I wanted to make a separate director repo and another one for all the local image repos (+ any other repo that has been authorized by the OEM and is kept safe). This way I can have the first check to see if the internal image repos are uncompromised. 
I use sets for this because of the unique element nature of the sets data structure, which would flag an error if the number exceeds 2, the number of fields we are checking for. 
The sets.difference() allows me to check if the director holds values which are the same as that of the other local/secure/offline imagerepos from the OEM.
And sets do all of this in constant time so I felt that this was the way to go. 

Do you still want me to change it?

shikhar394 committed:
Changes and some tests

shikhar394 committed:
Code Review changes made

awwad reviewed (commented):
Your commit summaries should be more descriptive in the future. "Changes and some tests" or "pushing changes" doesn't tell us anything. See [here for commit message guidelines](https://github.com/secure-systems-lab/lab-guidelines/blob/dev-workflow/commits.md#how-to-write-good-commit-messages). "Working Primary and demo_primary with manifests being submitted" is misleading, as they already were working. (Presumably, you mean that hardware_id and release_counter are now included in ECU or Vehicle manifests"?) Likewise for "Primary's clean_slate() works"

awwad commented on demo/demo_secondary.py:
```diff
@@ -499,10 +498,9 @@ def update_cycle():
   secondary_ecu.firmware_fileinfo = expected_target_info
-  secondary_ecu.update_release_counter(expected_target_info['fileinfo']['custom']['release_counter']) 
-  print("Updated", secondary_ecu)
+  secondary_ecu.update_release_counter(expected_target_info['fileinfo']\
```
FYI, the `\` here is not necessary because you're inside parentheses. You can exclude that.

Minor tip, also: if you're already using an additional line and the full contents of the parentheses will fit on that line, that's slightly easier to read than splitting the contents across two lines. (I think [PEP 8](https://www.python.org/dev/peps/pep-0008/) also recommends something similar, though it's a little more finicky about the details.)

So in this case:

```Python
  secondary_ecu.update_release_counter(
      expected_target_info['fileinfo']['custom']['release_counter'])
```

shikhar394 replied:
I'll make the changes. 
Thank you for pointing it out!!

awwad commented on uptane/clients/primary.py:
```diff
@@ -459,58 +455,63 @@ def get_validated_target_info(self, target_filepath):
 
+    director_target = validated_target_info[self.director_repo_name]
+    hardware_id_release_counter = set() # To check if the hardwareIDs and release counters are matching in all the repos other than director
+    director_hardware_id_release_counter = set() # To check with that of the other repos
```
I don't see the changes requested here:
- not distinguishing between Director and Image Repository in the code here
- using literal string and int rather than sets whose length you check

[Here's the full version of the resulting primary.py, with the section of code missing discussed changes highlighted](https://github.com/shikhar394/uptane/blob/86b826c7cdedf0b2c58ed9671a8984335f4b8cbe/uptane/clients/primary.py#L420-L484) at the time of this comment.

awwad commented:
(Also, don't stress about the gaps: it's expected that I'll find things, or we wouldn't need to bother with code reviews. Do try to check each comment afterwards to make sure it has been addressed, though, so the reviewer isn't going through again and catching the same issues. Note that Github will collapse the comments if the code they're attached to has been changed, but that doesn't mean the comment has been addressed; that makes it easier to miss some, so keep an eye out.)

I'll leave the rest alone until you look back through and let me know you think it's all addressed. When you're all done, I'll show you how I'd have done a few things and maybe that will be helpful, too.

shikhar394 committed:
Changed the primary.py file to reflect the algorithmic changes made towards finding
discrepencies in HardwareID and Release Counter in the image repositories/ director
repository.

shikhar394 committed:
Removed a comment that was written for explaining the process of updates in primary, which isn't yet set up.
Reference Link: https://github.com/uptane/uptane/pull/117#discussion_r129073727

shikhar394 committed:
Made suggested changes to the secondary.py *Removed import statements not required *Fixed formatting *Fixed if statements *Removed code that wasn't being used

shikhar394 committed:
Made code review changes to demo_secondary *Fixed spacing *Fixed longer lines as suggested

shikhar394 committed:
- Added support for exceptions tracking in primary. See self.primary_exceptions in the doc string.
- Made test cases to test fully_validate_metadata() in primary. See test_40 in test_primary.
- Removed the error class for BadHardwareIDReleaseCounter in __init__.py because realized that error will never be raised as either of the two will be caught and raised before the aforementioned error can be caught.
- Added a new folder with custom metadata for the new test cases. <samples/metadata_samples_long_expiry/updates_w_custom_field_to_one_ecu_bad_director> contains metadata manually produced to test the fully_validate_metadata function. Included images with mismatching hardwareIDs, mismatching release counters and a pristine image that can be downloaded successfully.

shikhar394 committed:
- Added docstring for primary
- Changed the format to 'der' in __init__.py

shikhar394 committed:
Removed print statements from primary.py

shikhar394 committed:
Made function for logging exceptions in primary

shikhar394 committed:
Working test_primary
Secondaries except errors raised in fully_validate_metadata
Sample metadata expire after 1 year
Test_secondary is still not working

shikhar394 committed:
Removed hardwareID and Release counter from primary

shikhar394 committed:
Removed more hardwareID and Release counter code from primary
